bind/bind.py

```diff
@@ -1,8 +1,26 @@
 """
-Bind: a one-line Python module to mimic os.path.join without reverting to root
+Bind: a Python module to mimic os.path.join without reverting to root
 """
 
 from urllib.parse import quote_plus
+from typing import Union, Optional
 
 
-def bind(*args, sep='/', url=True, safe=':/', **kwargs): return sep.join(quote_plus(str(arg).strip(sep), safe=safe, **kwargs) if url else str(arg).strip(sep) for arg in (args[0] if isinstance(args[0], (list, set, tuple)) else (i for i in args if i)))
+def bind(*args: Union[str, list, set, tuple],
+         sep: str = '/',
+         url: bool = True,
+         safe: str = ':/',
+         encoding: Optional[str] = None,
+         errors: Optional[str] = None) -> str:
+    if isinstance(args[0], (list, set, tuple)):
+        parts = args[0]
+    else:
+        parts = [i for i in args if i]
+
+    if url:
+        processed_parts = [quote_plus(str(arg).strip(sep), safe=safe, encoding=encoding, errors=errors)  # type: ignore
```

Comment:
This could be replaced with a generator expression instead of a list comprehension to save some memory, e.g.:
```
processed_parts = (quote_plus(str(arg).strip(sep), safe=safe, encoding=encoding, errors=errors) for arg in parts)
```
Same for the else branch below and for `parts` if type-hinting doesn't complain

Reply:
Thanks, made generators here. For `parts` typehinting expects same type as args, so its not happy with a generator as you suspected.